proposals/0072-feature-gate-threshold-automation.md

```diff
@@ -0,0 +1,193 @@
+
+### Step 1: Feature Creation
+
+The first step is creation of a feature account, done by submitting a
+transaction containing System instructions to fund, allocate, and assign the
+feature account to `Feature111111111111111111111111111111111111`.
+
+This step is unchanged from its original procedure.
+
+### Step 2: Staging Features for Activation
+
+A multi-signature authority, comprised of key-holders from Solana Labs and
+possibly other validator client teams in the future, will have the authority to
+stage created features for activation.
+In the future, this authority could be replaced by validator governance.
+
+The multi-signature authority stages a feature for activation by invoking the
+Feature Gate program's `StageFeatureForActivation` instruction. This instruction
+expects:
+
+- Data: The feature ID
+- Accounts:
+  - Staged Features PDA: writable
+  - Multi-signature authority: signer 
+
+One single PDA will be used to store two lists of features. Its data will be
+structured as follows:
+
+- 8 bytes for the current epoch.
+- A fixed-size list of size 8 containing 32-byte feature IDs staged for
+  activation at the end of the *current epoch*.
+- A fixed-size list of size 8 containing the `u64` value of total stake support
+  for each feature ID in the prior list.
+- 8 bytes for the next epoch.
+- A fixed-size list of size 8 containing 32-byte feature IDs staged for
+  activation at the end of the *next epoch*.
+
+`StageFeatureForActivation` will add the provided feature ID to the **next
+epoch's** set of staged features.
+
+An example of how this account's state might work to coordinate both lists can
+be found here:
+
+<https://github.com/buffalojoec/feature-gate-data-poc/blob/master/src/lib.rs>
+
+The Staged Features PDA could be derived simply from one literal seed:
+`"staged_features"`.
+
+### Step 3: Signaling Support for Staged Features
+
+With an on-chain reference point to determine the features staged for activation
+for a particular epoch, nodes can signal their support for the staged features
+supported by their software.
+
+A node signals its support for staged features by invoking the Feature Gate
+program's `SignalSupportForStagedFeatures` instruction. This instruction
+expects:
+
+- Data: A `u8` bit mask of the staged features.
+- Accounts:
+  - Staged Features PDA: writable
+  - Vote account: signer
+
+A `1` bit represents support for a feature. For example, for staged features
+`[A, B, C, D, E, F, G, H]`, if a node wishes to signal support for all features
+except `E` and `H`, their `u8` value would be 246, or `11110110`.
+
+A node's submitted bit mask is then used to add that node's current epoch stake
+to the Staged Features PDA for each feature for which it has signaled support.
+This is done by using the Get Epoch Stake Syscall.
+
+Nodes should send a transaction containing this instruction at some arbitrary
+point during the epoch at least 128 slots before the end of the epoch and on
+startup after any reboot.
```

Comment:
If a node does not send the transaction at all, is that tallied the same as if the node had sent a transaction indicating it does not support any of the features?

Reply:
Yes. If their transaction is either rejected (too late in the Epoch) or never sent, their stake support is never tallied. If they signal "no support" for all features, their stake is also not tallied.

Should this be included in the proposal?

Reply:
I think adding that reply to the SIMD would be great, just for completeness.